Test for OpenRPC subscription methods on trace/debug modules

ziogaschr opened the pull request:
Resolves https://github.com/etclabscore/core-geth/pull/345#discussion_r605569982

ziogaschr committed:
ethclient: remove duplicate rpc method name in tests

ziogaschr committed:
ethclient, node: test for subscription methods in openrpc discovery

ziogaschr committed:
ethclient: remove whitespace at EOL

ziogaschr committed:
node: add openrpc schema mapper for subscription IDs

ziogaschr committed:
node: add missing title for rpcSubscriptionParamsNameD schema mapper

ziogaschr committed:
node: add openrpc schema mappers for debug, trace subscription methods

ziogaschr commented on node/openrpc.go:
```diff
@@ -329,7 +369,8 @@ var blockNumberOrHashD = fmt.Sprintf(`{
 
-var rpcSubscriptionParamsNameD = `{
+var rpcEthSubscriptionParamsNameD = `{
+		"title": "subscriptionName",
```
@meowsbits added this, as it wasn't working for params without it even for `eth_subscribe`.

ziogaschr commented on node/openrpc.go:
```diff
@@ -208,25 +210,59 @@ func newOpenRPCDocument() *go_openrpc_reflect.Document {
 
 /*
-The following struct RPCSubscription and RPCSubscription.Unsubscribe method
+The following structs *RPCSubscription and *RPCSubscription.Unsubscribe method
```
@meowsbits I want to play a bit in order to find a better way of merging those 3 duplicated implementations where only the **subscriptionName** type in `subscribe` changes.

Let me know if you have an idea.

meowsbits replied:
Have you experimented at all with getting rid of or modifying the [general "skip PubSub" case](https://github.com/etclabscore/core-geth/blob/735fd8804dae5f612b7bddea3b441767784caa04/node/openrpc.go#L143) to be more permissive (even if only allowing `debug` and `trace` receivers through)?

ziogaschr replied:
No, I haven't. I will give it a try. Thanks

ziogaschr replied:
I have tried it here, which adds the `trace_filter` nicely, although it ~~can~~can't be used in the way documented, which leads to issues. What do you think? Did you mean something else @meowsbits?

![image](https://user-images.githubusercontent.com/802700/115776128-72322480-a3bc-11eb-9bc5-fd8d2bfdcc50.png)

meowsbits replied:
X-linking: https://github.com/etclabscore/core-geth/pull/364#pullrequestreview-643261538

ziogaschr commented on node/openrpc.go:
```diff
@@ -339,6 +380,20 @@ var rpcSubscriptionParamsNameD = `{
+var rpcDebugSubscriptionParamsNameD = `{
+		"title": "subscriptionName",
+		"oneOf": [
+			{"type": "string", "enum": ["traceFilter"], "description": "Returns traces for the filtered transactions within a range of blocks."}
```
I still have to collect all the `debug_*` subscription based methods

ziogaschr committed:
node: update openrpc schema mapper methods

ziogaschr committed:
node: whitelist skip pubsub to eth_ public api

I want to allow debug and trace API modules
through.

Taken from @meowsbits d7194b5

ziogaschr committed:
ethclient: generate docs for both `*_subscribe` as well Subscriptionable methods

ziogaschr committed:
docs/JSON-RPC-API/modules: run 'make docs-generate'

ziogaschr committed:
ethclient: remove js console example from generated docs for subscrionable methods

ziogaschr committed:
docs/JSON-RPC-API/modules: run 'make docs-generate'

meowsbits committed:
ethclient: fixup test again canonical methods list

Adapts the canonical list of methods that
should be provided by the document to meet
what the document says.

Date: 2021-04-30 06:13:54-05:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits reviewed (approved):
If this latest commit (https://github.com/etclabscore/core-geth/pull/356/commits/ce80270fa344ce41e6267e5e4a503604ea92be32) looks OK to you, everything LGTM! 

(I've left the methods commented because that lists serves a duty both as a test case, but also as a "behind-the-scenes" developer reference.)

ziogaschr committed:
node: openrpc to return the `*filters.PublicFilterAPI` methods

ziogaschr committed:
ethclient: generated docs have `*filters.PublicFilterAPI` methods as well as better examples

ziogaschr committed:
docs/JSON-RPC-API/modules: run 'make docs-generate'

ziogaschr committed:
ethclient: bring up tests against canonical methods list, including `*filters.PublicFilterAPI` methods
